File: src/RepoLite/Tests/RepoLite.VB.Tests/ActualGeneratedFileTests/GeneratedFiles/Base/BaseModel.vb
Imports System.Runtime.CompilerServices
Imports RepoLite.VB.Tests.REPOSITORYNAMESPACE.Base

Namespace MODELNAMESPACE.Base
    Public Class ValidationError
        Public Property PropertyName As String
        Public Property [Error] As String

        Public Sub New(ByVal [property] As String, ByVal [error] As String)
            PropertyName = [property]
            [Error] = [error]
        End Sub
    End Class

    Public Partial Interface IBaseModel
        ReadOnly Property EntityName As String
        Function SetValues(row As DataRow, propertyPrefix as String) As IBaseModel
    End Interface

    Public MustInherit Partial Class BaseModel
        Implements IBaseModel

        Public MustOverride ReadOnly Property EntityName As String Implements IBaseModel.EntityName
        Public MustOverride Function SetValues(row As DataRow, propertyPrefix as String) As IBaseModel Implements IBaseModel.SetValues
        Public MustOverride Function Validate() As List(Of ValidationError)
        Public ReadOnly DirtyColumns As List(Of String) = New List(Of String)()

        Public Sub ResetDirty()
            DirtyColumns.Clear()
        End Sub

        Protected Sub SetValue (Of T)(ByRef prop As T, ByVal value As T,
                                      <CallerMemberName> ByVal Optional propName As String = "")
            If Not DirtyColumns.Contains(propName) Then
                DirtyColumns.Add(propName)
            End If

            prop = value
        End Sub

        Public Shared Function GetDecimalPlaces(ByVal n As Decimal) As Integer
            n = Math.Abs(n)
            n -= CInt(n)
            Dim decimalPlaces = 0

            While n > 0
                decimalPlaces += 1
                n *= 10
                n -= CInt(n)
            End While

            Return decimalPlaces
        End Function
    End Class
End Namespace

File: src/RepoLite/RepoLite.Tests/GeneratedFiles/Models/Base/BaseModel.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Runtime.CompilerServices;

namespace NS.Models.Base
{
    public class ValidationError
    {
        public string PropertyName { get; set; }
        public string Error { get; set; }

        public ValidationError(string property, string error)
        {
            PropertyName = property;
            Error = error;
        }
    }
    
    public class ColumnDefinition
    {
        public string ColumnName { get; set; }
        public Type ValueType { get; set; }
        public string SqlDataTypeText { get; set; }
        public SqlDbType SqlDbType { get; set; }
        public bool Identity { get; set; }
        public bool PrimaryKey { get; set; }
        public bool Nullable { get; set; }

        public ColumnDefinition(string columnName) : this(columnName, typeof(string), "NVARCHAR(MAX)", SqlDbType.NVarChar, false, false, false) { }
        public ColumnDefinition(string columnName, Type valueType, string sqlDataTypeText, SqlDbType dbType) : this(columnName, valueType, sqlDataTypeText, dbType, false, false, false) { }
        public ColumnDefinition(string columnName, Type valueType, string sqlDataTypeText, SqlDbType dbType, bool nullable) : this(columnName, valueType, sqlDataTypeText, dbType, nullable, false, false) { }
        public ColumnDefinition(string columnName, Type valueType, string sqlDataTypeText, SqlDbType dbType, bool nullable, bool primaryKey) : this(columnName, valueType, sqlDataTypeText, dbType, nullable, primaryKey, false) { }
        public ColumnDefinition(string columnName, Type valueType, string sqlDataTypeText, SqlDbType dbType, bool nullable, bool primaryKey, bool identity)
        {
            ColumnName = columnName;
            ValueType = valueType;
            SqlDataTypeText = sqlDataTypeText;
            SqlDbType = dbType;
            Nullable = nullable;
            PrimaryKey = primaryKey;
            Identity = identity;
        }
    }
    
    public class TableDefinition
    {
        public string PrimaryKey { get; set; }
        public string Schema { get; set; }
        public string TableName { get; set; }
        public IEnumerable<ColumnDefinition> Columns { get; set; }
        public TableDefinition(string primaryKey, string schema, string tableName, IEnumerable<ColumnDefinition> columns)
        {
            PrimaryKey = primaryKey;
            Schema = schema;
            TableName = tableName;
            Columns = columns;
        }
    }
    
    public partial interface IBaseModel
    {
        string EntityName { get; }
        IBaseModel SetValues(DataRow row, string propertyPrefix);
    }

    public abstract partial class BaseModel : IBaseModel
    {
        public abstract string EntityName { get; }
        public abstract IBaseModel SetValues(DataRow row, string propertyPrefix);
        public abstract List<ValidationError> Validate();
        public readonly List<string> DirtyColumns = new List<string>();

        public void ResetDirty()
        {
            DirtyColumns.Clear();
        }

        protected void SetValue<T>(ref T prop, T value, [CallerMemberName] string propName = "")
        {
            if (!DirtyColumns.Contains(propName))
            {
                DirtyColumns.Add(propName);
            }
            prop = value;
        }

        public static int GetDecimalPlaces(decimal n)
        {
            n = Math.Abs(n);
            n -= (int)n;
            var decimalPlaces = 0;
            while (n > 0)
            {
                decimalPlaces++;
                n *= 10;
                n -= (int)n;
            }
            return decimalPlaces;
        }
    }
    
    public static class Ext
    {
        public static T? GetValue<T>(this DataRow row, string columnName) where T : struct
        {
            if (row.IsNull(columnName) || !row.Table.Columns.Contains(columnName))
                return null;
    
            return row[columnName] as T?;
        }
    
        public static string GetText(this DataRow row, string columnName)
        {
            if (row.IsNull(columnName) || !row.Table.Columns.Contains(columnName))
                return null;
    
            return row[columnName] as string;
        }
    }
}